File: aoc_2023/day8/day8.rb
require '../common'

LINES = getData('./input.txt', /\n/)

def hashAndStartBuilder(part2)
  start = ""
  hash = {}
  start_array = []
  LINES.each_with_index do |line, index|
    if index > 1
      array = line.split(" = ")
      key = array[0]
      if part2 == true
        start_array << key if key.end_with?('A')
      else
        start = key if key == "AAA"
      end
      coords = array[1].gsub(/[(),]/, "").split()
      hash[key] = {"L" => coords[0],
                    "R" => coords[1]
                  }
    end
  end
  if part2 == true
    return hash, start_array
  else
    return hash, start
  end
end

def solve_part1
  instructions = LINES[0].scan(/[RL]/)
  count = 0
  counter = 0
  length = instructions.length
  hash, start = hashAndStartBuilder(false)
  nextCoord = start

  while counter < length
    move = instructions[counter]
    nextCoord = hash[nextCoord][move]
    
    count += 1
    break if nextCoord == 'ZZZ'

    counter += 1
    if counter == (length)
      counter = 0
    end
  end
  return count
end


def solve_part2
  instructions = LINES[0].scan(/[RL]/)
  part2 = true
  start = ""

  hash, start_array = hashAndStartBuilder(part2)

  count = 0
  counter = 0
  length = instructions.length
  nextCoords = start_array
  loopTime = {}
    
  while counter < length
    move = instructions[counter]
    newNextCoords = []
    nextCoords.each_with_index do |coord, index|
      newCoord = hash[coord][move]
      if coord.end_with?("Z")
        loopTime[index] = count
      end
      newNextCoords << newCoord
    end
    nextCoords = newNextCoords
    count += 1
    counter += 1

    if counter == (length)
      counter = 0
    end

    if loopTime.values.length() == start_array.length
      return loopTime.values.reduce(1, :lcm)
    end
  end
end

p solve_part1

p solve_part2

File: aoc_2023/day8/day8.py
import re
from math import gcd

f = "test2.txt"
with open(f) as file:
  string = file.read().strip()
  lines = string.split('\n')


def compute_lcm(values):
   lcm = 1
   for value in values:
    lcm = (value*lcm)//gcd(value,lcm)
   return lcm


DICT = {'L': {}, 'R': {}}
instructions, coords = string.split('\n\n')
for line in coords.split('\n'):
  coord, direction = line.split(" = ")
  left,right = direction.split(',')
  left = left.strip()[1:].strip()
  right = right[:-1].strip()
  DICT['L'][coord] = left
  DICT['R'][coord] = right

def solve():
  startCoords = []
  # get starter coords - coords that end with 'A'
  for coord in DICT['L']:
    if coord.endswith("A"):
      startCoords.append(coord)

  TIMES = {}
  counter = 0
  while True:
    NEWPOINTERS = []
    for index,coord in enumerate(startCoords):
      # get coord for each instruction
      coord = DICT[instructions[counter%len(instructions)]][coord]
      if coord.endswith('Z'):
        TIMES[index] = counter+1
        #  have we captured all loop times? if so do lowest common multiple
        if len(TIMES) == len(startCoords):
          print(TIMES)
          return compute_lcm(TIMES.values())
      NEWPOINTERS.append(coord)
    #  progresses the coords down for the next loop
    startCoords = NEWPOINTERS
      

    counter += 1

print(solve())